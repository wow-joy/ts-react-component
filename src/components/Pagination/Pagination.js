import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import styled, { ThemeProvider } from "styled-components";
import Select from "../Select";
import { Type3 as Btn } from "../Btn";
import ControllSwitchHoc from "../../tools/Hoc/ControllSwitchHoc";

const Wrap = styled.div`
  display: flex;
  align-items: center;
  line-height: ${p => p.theme.size};
  font-size: 14px;
  & > * {
    flex-grow: 0;
    flex-shrink: 0;
    height: ${p => p.theme.size};
  }
  ${props => props.defaultStyles};
`;
const PageItem = styled.span`
  display: inline-block;
  padding: 0 8px;
  min-width: ${p => p.theme.size};
  text-align: center;
  cursor: pointer;
  &.active {
    color: #06aea6;
  }
  &.wjc-fast-jump__prev,
  &.wjc-fast-jump__next {
    position: relative;
    &::before {
      content: "···";
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      margin: auto;
      font-size: 18px;
      color: #999;
    }
  }
  &.wjc-fast-jump__prev:hover {
    position: relative;
    &::before {
      content: "<<";
      font-size: 12px;
      color: #06aea6;
      letter-spacing: -2px;
      transform: scale(0.8, 1.2);
    }
  }
  &.wjc-fast-jump__next:hover {
    position: relative;
    &::before {
      content: ">>";
      font-size: 12px;
      color: #06aea6;
      letter-spacing: -2px;
      transform: scale(0.8, 1.2);
    }
  }
`;
const Left = styled.div`
  width: ${p => p.theme.size};
  border: 1px solid #dbdbdb;
  cursor: pointer;
  position: relative;
  &.disable {
    cursor: not-allowed;
  }
  &::before {
    content: "";
    display: inline-block;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    width: 0;
    height: 0;
    margin: auto;
    border-right: 4px solid #ccc;
    border-top: 3px solid transparent;
    border-bottom: 3px solid transparent;
  }
`;
const Right = styled.div`
  width: ${p => p.theme.size};
  border: 1px solid #dbdbdb;
  cursor: pointer;
  position: relative;
  &.disable {
    cursor: not-allowed;
  }
  &::before {
    content: "";
    display: inline-block;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    width: 0;
    height: 0;
    margin: auto;
    border-left: 4px solid #ccc;
    border-top: 3px solid transparent;
    border-bottom: 3px solid transparent;
  }
`;
const Count = styled.div`
  margin: 0 12px;
`;
const SelectPageSize = styled(Select)`
  display: inline-block;
  width: auto;
  padding: 0 28px 0 12px;
  &::after {
    right: 10px;
  }
  li:hover {
    background: #fffbe0;
  }
  li.active {
    color: #06aea6;
  }
`;
const JumpTo = styled.div`
  display: flex;
  align-items: center;
  margin: 0 10px;
  & > input[type="number"] {
    height: ${p => p.theme.size};
    background: transparent;
    width: ${p => p.theme.size};
    text-align: center;
    border: 1px solid #dbdbdb;
    margin: 0 7px;
    -moz-appearance: textfield;
    &::-webkit-outer-spin-button {
      -webkit-appearance: none;
    }
    &::-webkit-inner-spin-button {
      -webkit-appearance: none;
    }
  }
`;
const Submit = styled(Btn)`
  line-height: ${p => p.theme.size};
`;
class Pagination extends PureComponent {
  state = {
    currentPage: "",
    jumpToValue: ""
    // pageSize: this.props.defaultPageSize
  };

  componentWillMount() {
    if (!this.props.pageSize) {
      throw new Error("please set pageSize, 请设置pageSize 或 defaultPageSize");
    }
    if (!this.props.pageSizeList) {
      throw new Error("please set pageSizeList, 请设置pageSizeList");
    }
  }

  render() {
    const {
      className,
      defaultStyles,
      viewAble = [
        "prevNext",
        "pageList",
        "total",
        "pageSizeSelect",
        "jumpTo",
        "submit"
      ],
      staticStr = ["共", "条", "条/页", "跳至", "页", "确定"],
      currentPage = 1,
      siblingViewSize = 2,
      size = "32px",
      total,
      pageSizeList
    } = this.props;
    const { pageSize } = this.props;
    // const currentPage = value;
    const pageLength = Math.ceil(total / pageSize);
    const pageArr = this.getPageArr(pageLength, siblingViewSize, currentPage);
    return (
      <ThemeProvider theme={{ size: size }}>
        <Wrap
          defaultStyles={defaultStyles}
          className={"wjc-pagination " + className || ""}
        >
          {viewAble.includes("prevNext") && (
            <Left
              onClick={currentPage === 1 ? null : this.goto(currentPage - 1)}
              className={`wjc-page-prev ${currentPage === 1 ? "disable" : ""}`}
            />
          )}
          {viewAble.includes("pageList") &&
            pageArr.map(ele => {
              if (typeof ele == "number") {
                return (
                  <PageItem
                    key={ele}
                    onClick={this.goto(ele)}
                    className={`wjc-page-item ${
                      ele === currentPage ? "active" : ""
                    }`}
                  >
                    {ele}
                  </PageItem>
                );
              }
              return (
                <PageItem
                  key={ele}
                  onClick={this.goto(
                    ele === "prev"
                      ? currentPage - siblingViewSize * 2 - 1
                      : currentPage + siblingViewSize * 2 + 1
                  )}
                  title={
                    ele === "prev"
                      ? `前进${siblingViewSize * 2 + 1}页`
                      : `向后${siblingViewSize * 2 + 1}页`
                  }
                  className={`wjc-fast-jump__${ele}`}
                />
              );
            })}
          {viewAble.includes("prevNext") && (
            <Right
              onClick={
                currentPage === pageLength ? null : this.goto(currentPage + 1)
              }
              className={`wjc-page-next ${
                currentPage === pageLength ? "disable" : ""
              }`}
            />
          )}
          {viewAble.includes("total") && (
            <Count className={"wjc-page-count"}>{`${staticStr[0]}${total}${
              staticStr[1]
            }`}</Count>
          )}
          {viewAble.includes("pageSizeSelect") && (
            <SelectPageSize
              className={"wjc-page-size__select"}
              value={pageSize}
              inputRender={({ value }) => (value ? value.label : "")}
              options={pageSizeList.map(ele => ({
                label: `${ele}${staticStr[2]}`,
                value: ele
              }))}
              onChange={this.changePageSize}
            />
          )}
          {viewAble.includes("jumpTo") && (
            <JumpTo className={"wjc-jump-to"}>
              {staticStr[3]}
              <input
                type="number"
                onKeyDown={this.keyDownJumpTo}
                value={this.state.jumpToValue}
                onChange={this.changeJumpToInput}
              />
              {staticStr[4]}
            </JumpTo>
          )}
          {viewAble.includes("submit") && (
            <Submit
              className={"wjc-jump-to__submit"}
              onClick={this.submitJumpTo}
            >
              {staticStr[5]}
            </Submit>
          )}
        </Wrap>
      </ThemeProvider>
    );
  }
  goto = page => e => {
    const { onChange } = this.props;
    onChange && onChange(page, this.props.pageSize, this.props.total);
  };
  getPageArr = (pageLength, siblingViewSize, currentPage) => {
    const centerVisibleSize = siblingViewSize * 2 + 1;
    if (pageLength <= centerVisibleSize + 4) {
      // 足够显示所有项目时不显示`...`快速跳转
      return Array(pageLength)
        .fill("")
        .map((ele, index) => index + 1);
    } else {
      const centerArr = [];
      const fillArr = (min, max) => {
        return Array(max - min + 1)
          .fill("")
          .map((ele, index) => min + index);
      };

      if (currentPage <= 2 + siblingViewSize) {
        let arr = [
          ...fillArr(
            1,
            Math.max(centerVisibleSize, currentPage + siblingViewSize)
          ),
          "next",
          pageLength
        ];
        return arr;
      }
      if (currentPage >= pageLength - siblingViewSize - 1) {
        let arr = [
          1,
          "prev",
          ...fillArr(
            Math.min(
              pageLength - centerVisibleSize + 1,
              currentPage - siblingViewSize
            ),
            pageLength
          )
        ];
        return arr;
      }
      return [
        1,
        "prev",
        ...fillArr(currentPage - 2, currentPage + 2),
        "next",
        pageLength
      ];
    }
  };
  changePageSize = value => {
    const { onPageSizeChange, onChange } = this.props;
    onChange && onChange(this.props.value || 1, value, this.props.total);
    onPageSizeChange && onPageSizeChange(value);
    // this.setState({
    //   pageSize: value
    // });
  };
  changeJumpToInput = e => {
    this.setState({
      jumpToValue: e.target.value
    });
  };
  keyDownJumpTo = e => {
    if (e.keyCode === 13) {
      this.submitJumpTo();
    }
    if (e.keyCode === 27) {
      this.clearJumpTo();
    }
  };
  submitJumpTo = () => {
    if (this.state.jumpToValue === "") {
      return false;
    }
    const { total } = this.props;
    const { pageSize } = this.props;
    const pageLength = Math.ceil(total / pageSize);
    this.goto(Math.min(Math.max(this.state.jumpToValue, 1), pageLength))();
    this.clearJumpTo();
  };
  clearJumpTo = () => {
    this.setState({
      jumpToValue: ""
    });
  };
}

Pagination.propTypes = {
  className: PropTypes.string,
  defaultStyles: PropTypes.string,
  size: PropTypes.string,
  viewAble: PropTypes.array,
  staticStr: PropTypes.array,
  total: PropTypes.number,
  pageSize: PropTypes.number,
  onChange: PropTypes.func,
  onPageSizeChange: PropTypes.func,
  siblingViewSize: PropTypes.number,
  currentPage: PropTypes.number,
  pageSizeList: PropTypes.array
};

export default ControllSwitchHoc({
  value: "pageSize",
  onChange: "onPageSizeChange",
  defaultValue: "defaultPageSize"
})(
  ControllSwitchHoc({
    value: "currentPage",
    defaultValue: "defaultCurrentPage"
  })(Pagination)
);
