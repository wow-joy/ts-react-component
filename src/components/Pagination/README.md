# Pagination

Pagination 组件

## 使用

- 基础用法

```
import React from 'react'
import { Pagination } from 'wowjoy-component'

const PROPS = {
  size: "32px",
  total: 1000,
  pageSizeList: [10, 20, 30],
  defaultPageSize: 10,
  onChange: (...args)=>console.log(args)
};
const Foo = () => <Pagination {...PROPS}></Pagination>
```

## APIs

| 属性               | 描述                                                                      |  类型  |                                 默认值                                  |
| ------------------ | :------------------------------------------------------------------------ | :----: | :---------------------------------------------------------------------: |
| className          | 顶层样式 class                                                            | string |                                                                         |
| defaultStyles      | 顶层默认样式                                                              | string |                                                                         |
| children           | 内容                                                                      |  node  |                                                                         |
| viewAble           | 可见项配置; 请从默认值中删除不要字段来隐藏组件，不要修改字段              |  arr   | ['prevNext', 'pageList', 'total', 'pageSizeSelect', 'jumpTo', 'submit'] |
| staticStr          | 静态文本配置; 请从默认值中修改字段，不要改变顺序或删除                    |  arr   |               ['共', '条', '条/页', '跳至', '页', '确定']               |
| size               | 尺寸                                                                      | string |                                 '32px'                                  |
| total              | 总数，<b>必填项</b>                                                       | number |                                                                         |
| pageSize           | 每页数量， 受控  <b>pageSize,defaultPageSize 必须填写其中之一</b>         | number |                                                                       |
| defaultPageSize    | 初始每页数量， 非受控                                          | number |                                                                         |
| onPageSizeChange   | 每页数量（pageSize）改变事件<br/>( pageSize ) =>{} |  func  |                                                                        |
| pageSizeList       | 每页数量可惜列表，<b>必填项</b>                                           | arr |                                                                         |
| siblingViewSize    | 两侧可见页数量，如传入 2，则中间会显示 2+1+2 页                           | number |                                    2                                    |
| currentPage        | 当前所在页， 受控                                                         | number |                                    1                                    |
| defaultCurrentPage | 初始化当前所在页， 非受控                                                 | number |                                    1                                    |
| onChange           | 当前页改变事件，pageSize 改变也会触发本事件<br/>(currentPage, pageSize, total) =>{} |  func  |                                                                        |

## classApi

| class                 | 描述             |
| --------------------- | :--------------- |
| wjc-page-prev          | 上一页           |
| wjc-page-next          | 下一页           |
| wjc-page-item(.active) | 页码(当前页)     |
| wjc-fast-jump\_\_prev  | 左侧快速跳转     |
| wjc-fast-jump\_\_next  | 右侧快速跳转     |
| wjc-page-count         | 总页数           |
| wjc-jump-to            | 快速跳转         |
| wjc-jump-to\_\_submit  | 快速跳转确认按钮 |
