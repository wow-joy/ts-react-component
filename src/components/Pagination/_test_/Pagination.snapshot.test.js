import React from "react";
import Pagination from "../index.js";
import renderer from "react-test-renderer";
import * as TestUtils from "react-dom/test-utils";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

describe("Pagination", () => {
  const reactObj = (
    <Pagination pageSizeList={[10, 20, 30]} total={100} pageSize={10} />
  );
  it("renders correctly", () => {
    const tree = renderer.create(reactObj).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
