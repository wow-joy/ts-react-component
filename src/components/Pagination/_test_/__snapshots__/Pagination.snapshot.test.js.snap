// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div
  className="wjc-pagination undefined sc-EHOje gNDNVt"
>
  <div
    className="wjc-page-prev disable sc-gzVnrw dydKHp"
    onClick={null}
  />
  <span
    className="wjc-page-item active sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    1
  </span>
  <span
    className="wjc-page-item  sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    2
  </span>
  <span
    className="wjc-page-item  sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    3
  </span>
  <span
    className="wjc-page-item  sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    4
  </span>
  <span
    className="wjc-page-item  sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    5
  </span>
  <span
    className="wjc-fast-jump__next sc-bZQynM cQoCtO"
    onClick={[Function]}
    title="向后5页"
  />
  <span
    className="wjc-page-item  sc-bZQynM cQoCtO"
    onClick={[Function]}
  >
    10
  </span>
  <div
    className="wjc-page-next  sc-htoDjs fDYsde"
    onClick={[Function]}
  />
  <div
    className="wjc-page-count sc-dnqmqq cHIqGk"
  >
    共100条
  </div>
  <div
    className="wjc-select wjc-page-size__select sc-iwsKbI bgcDVW  sc-bdVaJa iwzsiR"
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH fNOhzS"
    >
      10条/页
    </div>
    <ul
      className="wjc-select-list sc-htpNat caZwnp"
      onAnimationEnd={[Function]}
      onClick={[Function]}
    >
      <li
        className="active sc-bxivhb cdHdp"
        onClick={[Function]}
      >
        10条/页
      </li>
      <li
        className="sc-bxivhb cdHdp"
        onClick={[Function]}
      >
        20条/页
      </li>
      <li
        className="sc-bxivhb cdHdp"
        onClick={[Function]}
      >
        30条/页
      </li>
    </ul>
  </div>
  <div
    className="wjc-jump-to sc-gZMcBi hXGulT"
  >
    跳至
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      type="number"
      value=""
    />
    页
  </div>
  <span
    className="wjc-btn  wjc-jump-to__submit sc-gqjmRU gAsrvk sc-ifAKCX dmsRxv"
    onClick={[Function]}
  >
    确定
  </span>
</div>
`;
